internal/filesystem/filesystem.go

```diff
@@ -1,31 +1,164 @@
 package filesystem
 
 import (
+	"context"
+	"errors"
+	"fmt"
 	"io"
 	"io/fs"
+	"os"
+	"time"
 )
 
-// TODO: name?
-type FS interface {
+type (
+	// TODO: [review] These interfaces come from our pre fs.FS implementations
+	// Things like names and patterns may (need to) change.
+	// Some things might need to be added as well.
+	IDFS interface {
+		fs.FS
+		ID() ID
+	}
+	// TODO: better name; OpenPOSIXFile?
+	OpenFileFS interface {
+		fs.FS
+		OpenFile(name string, flag int, perm fs.FileMode) (fs.File, error)
+	}
+	MakeFileFS interface {
+		fs.FS
+		MakeFile(string) error
+	}
+	RemoveFS interface {
+		fs.FS
+		Remove(string) error
+	}
+	SymlinkFS interface {
+		fs.FS
+		// TODO: better name?
+		// Probably should be explicit to avoid namespace collisions
+		// with hard-linkers.
+		MakeLink(string, string) error
+		// TODO: This is legacy and should be changed
+		// We used to collide with "ReadLink".
+		// We can use that name, or something else now.
+		ExtractLink(string) (string, error)
+	}
+	RenameFS interface {
+		fs.FS
+		// TODO: We need to document expectations for this.
+		// Specifically what error value is expected to be returned
+		// if things like type constraints are enforced for the system.
+		// I.e. "... should return [ErrX] if old is type X
+		// and new is type Y and the system disallows this"
+		// ^ But in English [Ame].
```

Comment:
Let's ignore this for now.
It's not clear if following the pre-1.13 standard and demanding `os.LinkError` is a good idea, although `fs` uses `os|fs.PathError`.
We can make a standards/documentation pass for our extension interfaces, after our own implementations of them see some usage.
https://github.com/djdv/go-filesystem-utils/issues/22 will probably factor into this somehow as well.

Reply:
Remark removed in f9a75020973f3af2b73f2889cb025e6c094904a8